llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7067,6 +7067,10 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
         SrcVecTy = smallestIntegerVectorType(SrcVecTy, MinVecTy);
         VectorTy =
             largestIntegerVectorType(ToVectorTy(I->getType(), VF), MinVecTy);
```

Comment:
I do think the logic here is wrong: we do truncation from largest to smallest, so computation of `SrcVecTy` and `VectorTy` should be swapped.
Looks like so far this code was "lucky" as non-RISC-V TTI tolerates such discrepancy between src and dst types.

Reply:
Looks like it is intended in the original patch 55d633bd60235.
Think about this for a while but I haven't fully grasped the author's intentions yet.

Reply:
Even if `canTruncateToMinimalBitwidth` is true, the destination type should be no smaller than source type for `TruncInst`. The other easiest way to verify that `SrcVecTy` and `VectorTy` should be swapped is by running existing littest with assert when  `SrcVecTy > VectorTy`: 
```
llvm/test/Transforms/LoopVectorize/trunc-reductions.ll:

opt -passes=loop-vectorize,dce,instcombine -force-vector-interleave=1 -force-vector-width=8

Trunc: SrcVecTy = <8 x i1>      VectorTy = <8 x i16>
```
If predicate is reversed, there is not going to fail.
Other part of the code: VPlan construction and generation are done correctly, so no wonder it was not discovered before.

Reply:
@arcbbb @fhahn